src/mtree_core.erl

```diff
@@ -0,0 +1,235 @@
+%% @doc This module contatins helpers function for the merkle tree API
+%% functions.
+%% @end
+-module(mtree_core).
+
+-include("ppspp.hrl"). 
+
+-export([hash/1,
+         compare_hash/2,
+         lies_in/2,
+         get_parent/1,
+         get_sibling/1,
+         get_layer_num/1,
+         bin_to_range/1,
+         pad_tree/1,
+         nearest_power_2/1,
+         is_complete/1,
+         root_bin/1,
+         get_munro_root/1,
+         get_next_munro_root/1,
+         next_bin/1,
+         tree_length/1]).
+
+-opaque mtree()     :: atom().
+-type bin()         :: non_neg_integer().
+-type hash()        :: binary().
+-type hash_list()   :: list(hash()).
+
+-export_type([mtree/0]).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc returns hash of Hash_List
+%% @end
+-spec hash(hash_list()) -> hash().
+hash(Hash_List) ->
+    crypto:hash(?PPSPP_DEFAULT_HASH_ALGORITHM, Hash_List).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc returns hash of Hash_List
+%% @end
+-spec compare_hash(hash(), hash()) -> true | false.
+compare_hash(H, H) -> true;
+compare_hash(_, _) -> false.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc lies_in/2 : return true if the bin lies in the given range.
+%% @end
+-spec lies_in(bin(), list())-> true |
+                               false. 
+lies_in(Bin, [Start, End]) when Bin >= Start andalso Bin =< End ->
+    true;
+lies_in(_, _) ->
+    false.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc returns the sibling bin number for a bin number belonging to a given
+%% layer
+%% @end
+-spec get_parent(bin()) -> bin().
+get_parent(Bin) ->
+    erlang:round((get_sibling(Bin) + Bin)/2).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc returns the sibling bin number for a bin number belonging to a given
+%% layer
+%% @end
+-spec get_sibling(bin()) -> bin().
+get_sibling(Bin) ->
+    Layer = get_layer_num(Bin)-1,
+    Nth   = erlang:round((Bin-math:pow(2,Layer)+1)/math:pow(2,Layer+1))+1,
+    if
+        Nth rem 2 =:= 0 -> Bin - erlang:round(math:pow(2,Layer+1));
+        true            -> Bin + erlang:round(math:pow(2,Layer+1))
+    end.
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% @doc Return the layer number to which a particular bin belongs.
+%% @end
+-spec get_layer_num(bin()) -> pos_integer().
+get_layer_num(Bin) when Bin >= 0 ->
+    [Start, End] = bin_to_range(Bin),
+    erlang:round(math:log(End+2 -Start)/math:log(2)).
+
+%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+%% TODO add check for Bin to be an integer in the gaurd
+%% @doc Return [Start, End] range of leaf nodes for a given Bin
+%% @end
+-spec bin_to_range(bin()) -> list(bin()).
+bin_to_range(Bin) when Bin >= 0 ->
+    if
```

Comment:
Replace the 'if' expression on line 106 with a 'case' expression or function clauses.